specs/registry-auth.md

```diff
@@ -0,0 +1,143 @@
+       reg -- Response                --> client;
+```
+
+### Token Scheme
+
+Notation supports two types of token schemes, [Docker][token] and [OAuth 2.0][RFC6749] where OAuth 2.0 is preferred by default.
+
+Since token authentication schemes have two more requests per registry requests, notation client implementations SHOULD cache the token to reduce the number of requests. 
+
+```mermaid
+flowchart LR;
+    ts[Authorization Service];
+    client[Notation];
+    reg[Registry];
+
+    client -- 1. Access resource with cached token --> reg;
+       reg -- 2. Challenge if token invalidated    --> client;
+    client -- 3. Request for a new token           --> ts;
+        ts -- 4. Grant a new token                 --> client;
+    client -- 5. Access with a new token           --> reg;
+       reg -- 6. Response                          --> client;
+```
+
+The workflow is updated as follows with caching.
+
+1. Notation attempts to access the remote registry using a cached token. If the token is valid, skip to *step 6*.
+2. The remote registry returns `401 Unauthorized` with a challenge, requiring a valid token.
+3. Notation requests the authorization service for a new bearer token for accessing the target resource with local credentials.
+4. The authorization grants and returns a new bearer token for access back to the notation client. The client-side cache is refreshed by the newly returned token.
+5. Notation attempts to access the remote registry again.
+6. The remote registry performs the requested operation and returns the response.
+
+#### Docker
+
+In the [Docker Token Authentication][token] specification, the *step 3* is implemented using a `GET` request where users are authenticated by the authorization service via `Basic` authentication scheme. Therefore, user credentials are only accepted in the form of username and password pair.
+
+#### OAuth 2
+
+Notation follows the [Docker Registry v2 authentication][oauth2] specification for the [OAuth 2.0][RFC6749] framework where the *step 3* is implemented using a `POST` request. Precisely, notation supports `password` and `refresh_token` grant types.
+
+**Note** Refresh tokens are often known as identity tokens in the context of registry authentication.
+
+
+## Credential Store
+
+As local credentials may be required to access the remote registries, they need to be stored and accessed securely. To achieve maximum security, credential helpers are preferred so that credentials are stored in the system key chain with better protection. If credential helpers are not available, notation will fall back to credential files with proper access control.
+
+### Credential Helper
+
+To achieve maximum compatibility with existing systems, [docker credential helpers](https://github.com/docker/docker-credential-helpers) and its [protocol](https://docs.docker.com/engine/reference/commandline/login/#credential-helper-protocol) are adopted as the credential helpers for `notation`.

```

Comment:
Which [cred helper](https://github.com/docker/docker-credential-helpers#development) commands will notation use, only `get`?

Reply:
`get` for registry access, and `store` for `notation login` (see #119)